keps/sig-storage/3314-csi-changed-block-tracking/README.md

```diff
@@ -0,0 +1,1549 @@
+by creating a [SnapshotMetadataService CR](#snapshot-metadata-service-custom-resource)
+that contains the service's TCP endpoint address, CA certificate and
+an audience string needed for token authentication.
+The CSI driver name is specified in a metadata label in this
+CR, so that a backup application can efficiently search for
+the [SnapshotMetadata](#the-snapshotmetadata-service-api) gRPC service
+of the provisioner of the VolumeSnapshots to be backed up.
+
+Before using the [SnapshotMetadata](#the-snapshotmetadata-service-api) gRPC API,
+a backup application must first obtain an authentication token using the Kubernetes
+[TokenRequest API](https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/token-request-v1/)
+with the service's audience string.
+It should establish trust with the specified CA for use in gRPC calls and
+then directly make TLS gRPC calls to the
+[SnapshotMetadata](#the-snapshotmetadata-service-api) gRPC service's
+TCP endpoint.
+The audience-scoped authentication token must be passed in the metadata
+of each RPC as the value of the ``authorization`` metadata key;
+it will be used to authorize the backup application's use of the service.
+Every RPC returns a gRPC stream through which the metadata can be recovered.
+
+The advertised [SnapshotMetadata](#the-snapshotmetadata-service-api) gRPC service's
+TCP endpoint need not actually be hosted by CSI driver vendor supplied logic,
+but could be from a new community provided
+[external-snapshot-metadata sidecar](#the-external-snapshot-metadata-sidecar)
+that communicates over a private UNIX domain socket with the CSI driver's
+implementation of the [SnapshotMetadata](#the-snapshotmetadata-service-api)
+gRPC service.
+Such a CSI driver service need only focus on the generation of the metadata
+requested;
+the sidecar is responsible for validating the authentication token,
+authorizing the backup application and validating the parameters of the RPC calls.
+The sidecar forwards the RPC call to the CSI driver service over the UNIX domain socket,
+after translating Kubernetes object names into SP object names,
+and re-streams the results back to its client.
+
+The process of accessing snapshot metadata via the sidecar is illustrated in the figure below.
+Additional information is available in the [Design Details](#design-details) section.
+
+![metadata retrieval flow](./flow.drawio.svg)
+
+### User Stories
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+
+#### Full snapshot backup
+
+A backup application needs to perform a full backup on volumes of a specific
+Kubernetes application.
+
+For each volume in the application:
+1. The backup application creates a VolumeSnapshot of a PVC that needs to be
+backed up.
```

Comment:
Is it expected that the backup application creates the snapshot of the PVC? Assuming that the namespace where the backup application runs will usually be different from the namespace where the application-to-backup is running.

Could (user) application namespaces have a cronjob like setup to create snapshots at certain times/intervals? Or is it expected that backup applications offer mechanisms for users to configure this?

Possibly this is all a non-goal, and can be added in the appropriate section.

Reply:
Yeah, the CBT snapshot metadata service just assumes that the snapshots have been created beforehand. The actual snapshotting mechanism used by the backup application (and the RBAC involves) is out of the scope of this KEP, as long as the snapshots are CSI snapshots.

Generally, backup applications would create the snapshots in the target namespace (and the underlying cluster-scoped snapshot content), and run the backup on PVCs restored from those snapshots to ensure crash consistency.

Reply:
Sounds good to me. It may be nice to include such a description in the KEP as well.